### PR TITLE
Warn about multiple conflicting #[repr] hints

### LINKED ISSUE #34622: #[repr(u64)], #[repr(C, u64)], #[repr(u64, C)] have different behavior
# Code

```rust
use std::mem;

#[repr(u64)]
enum BookHorses { TwilightSparkle }

#[repr(C, u64)]
enum PurpleHorses { TwilightSparkle }

#[repr(u64, C)]
enum PrincessHorses { TwilightSparkle }

fn main() {
    println!("{}", mem::size_of::<BookHorses>());
    println!("{}", mem::size_of::<PurpleHorses>());
    println!("{}", mem::size_of::<PrincessHorses>());
}
```

# Output

```
8
4
8
```

i.e. when the compiler sees multiple `repr` arguments, it takes the first one and ignores the rest.

# Expected behavior

Either all enums should have the same size, or the latter enums should not compile.

### PR DESCRIPTION
**Pull request by lambda-fairy**
Closes #34622

### CODE REVIEW

**Comment by Aatch** [comment]
I wonder if this should instead be a warning instead of an error? Also, @nagisa is the use of `#[repr(C, u32)]` in the failing test deliberate?

**Comment by nagisa** [comment]
@Aatch what do you mean/are you referring to?

**Comment by Aatch** [comment]
@nagisa one of the tests is failing, `mir_adt_construction.rs` because it uses `#[repr(C, u32)]`. You wrote the test, so I pinged you about it.

**Comment by lfairy** [comment]
Okay, fixed the error in `mir_adt_construction.rs`. Lets see if Travis likes it now.

**Comment by nagisa** [comment]
Ah, no, I don’t think it was intentional.

**Comment by lfairy** [comment]
Okay, I've switched to a warning instead.

Should we do a crater run with this patch? I'm curious as to which crates make this mistake in the wild.

**Comment by lfairy** [comment]
Just noticed this weird test failure:

```
failures:

---- [pretty] pretty/conflicting-repr-hints.rs stdout ----
	
error: pretty-printed source does not match expected source

expected:
------------------------------------------
// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#![allow(dead_code)]

#[repr(C)]
enum A { A, }

#[repr(u64)]
enum B { B, }

 //~ WARNING conflicting representation hints
#[repr(C, u64)]
enum C { C, }

 //~ WARNING conflicting representation hints
#[repr(u32, u64)]
enum D { D, }

#[repr(C, packed)]
struct E(i32);

fn main() { }

------------------------------------------
actual:
------------------------------------------
// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#![allow(dead_code)]

#[repr(C)]
enum A { A, }

#[repr(u64)]
enum B { B, }

//~ WARNING conflicting representation hints
#[repr(C, u64)]
enum C { C, }

//~ WARNING conflicting representation hints
#[repr(u32, u64)]
enum D { D, }

#[repr(C, packed)]
struct E(i32);

fn main() { }

------------------------------------------
```

Anyone know what's going on here? The test seems to expect an extra space (??)

**Comment by lfairy** [comment]
I've rewritten the patch to use the HIR machinery. It should be ready to merge now.

Thanks @eddyb for your help on IRC!

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -1726,6 +1726,7 @@ fn cookie() -> ! { // error: definition of an unknown language item: `cookie`
 
 register_diagnostics! {
 //  E0006 // merged with E0005
+    E0011, // conflicting representation hints
```
Comment (eddyb) [inline src/librustc/diagnostics.rs]: You should pick the first available number *after* the highest in use.

**Comment by lfairy** [comment]
@eddyb I've changed the error code to the one after the last.

Should I squash the two commits, or leave them as-is?

**Comment by eddyb** [comment]
@lfairy Squashing/amending is preferred in such cases.

**Comment by lfairy** [comment]
Sure, I've squashed it now.

**Comment by lfairy** [comment]
Rebased.

**Comment by lfairy** [comment]
Rebased again.

By the way: is anyone looking at this PR? It's been open for a while now, and I'd rather get it merged soon before the diagnostics system gets more churn.

**Comment by lfairy** [comment]
Apparently someone else added a diagnostic code in the mean time 😢 Should be fixed now.

@Aatch re-r?

**Pull request merged**